qabel_provider/models.py

```diff
@@ -64,6 +87,120 @@ def send_confirmation_mail(self):
         EmailAddress.objects.get_primary(self.user).send_confirmation(signup=False)
 
 
+class PlanInterval(models.Model, ExportModelOperationsMixin('planinterval')):
+    # A user may own any number of (presumably prepaid) plan intervals
+    profile = models.ForeignKey(Profile)
+    plan = models.ForeignKey(Plan)
+
+    # Duration the interval can be used
+    duration = models.DurationField()
+
+    STATES = (
+        ('pristine', 'pristine (neither expired nor used)'),
+        ('in_use', 'currently used interval'),
+        # expired := now() past (started_at + duration)
+        ('expired', 'expired interval'),
+    )
+    state = models.CharField(choices=STATES, default='pristine', max_length=10)
+
+    # Use started at <timestamp>
+    started_at = models.DateTimeField(null=True, blank=True)
+
+    def check_expiry(self):
+        """
+        Return false-ish if plan is expired, self otherwise.
+
+        Update state if expired.
+        """
+        if timezone.now() > (self.started_at + self.duration):
+            self.state = 'expired'
+            audit_log = ProfilePlanLog(profile=self.profile,
+                                       action='expired-interval', plan=self.plan, interval=self)
+            with transaction.atomic():
+                self.save()
+                audit_log.save()
+            return
+        return self
+
+    def start(self):
+        """Start using this (pristine) interval."""
+        if self.state != 'pristine':
+            raise ValueError('Cannot start using a %s interval' % self.state)
+        self.state = 'in_use'
+        self.started_at = datetime.datetime.now()
+        audit_log = ProfilePlanLog(profile=self.profile,
+                                   action='start-interval', plan=self.plan, interval=self)
+        with transaction.atomic():
+            self.save()
+            audit_log.save()
+
+    @classmethod
+    def get_or_start_interval(model, profile):
+        """Return/activate a plan interval for *profile*, or None."""
+        with transaction.atomic():
+            interval = model._get_interval(profile)
+            if not interval:
+                interval = model._start_interval(profile)
+            return interval
+
+    @classmethod
+    def _get_interval(model, profile):
+        """Return plan interval used by *profile*, or None."""
+        try:
+            interval = model.objects.get(profile=profile, state='in_use')
+        except ObjectDoesNotExist:
+            return
+        else:
+            return interval.check_expiry()
+
+    @classmethod
+    def _start_interval(model, profile):
+        """Begin new plan interval for *profile*, or None if no intervals are available (anymore)."""
+        usable_interval = model.objects.filter(profile=profile, state='pristine').first()
+        if not usable_interval:
+            return
+        usable_interval.start()
+        return usable_interval
```

Comment:
while the above code is covered via the REST tests it should also have some unit tests